packages/tailwind/src/tailwind.tsx

```diff
@@ -19,214 +12,127 @@ export interface TailwindProps {
+          a user-defined class could end up also being sanitized which would lead to unexpected 
+          behavior and bugs that are hard to track.
+        */
+        for (const singleClass of nonInlinableClasses) {
+          propsToOverwrite.className = propsToOverwrite.className.replace(
+            singleClass,
+            sanitizeClassName(singleClass),
+          );
+        }
+      } else {
+        propsToOverwrite.className = undefined;
+      }
+    }
 
-  let nonMediaQueryCSS = markupCSS;
-  const nonEscapedMediaQueryClasses = [] as string[];
-
-  for (const [mediaQuery, content] of markupCSS.matchAll(
-    /@media\s*\(.*\)\s*{(\s*\..*\s*{[\s\S]*}\s*)}/gm,
-  )) {
-    nonMediaQueryCSS = nonMediaQueryCSS.replace(mediaQuery, "");
-
-    let finalMediaQuery = mediaQuery;
-
-    for (const [
-      fullRule,
-      escapedRuleClassName,
-      ruleContent,
-    ] of content.matchAll(/\s*\.([\S]+)\s*{([^}]*)}/gm)) {
-      const ruleClassName = escapedRuleClassName.replaceAll(/\\[0-9]|\\/g, "");
-      nonEscapedMediaQueryClasses.push(ruleClassName);
-
-      finalMediaQuery = finalMediaQuery.replace(
-        fullRule,
-        fullRule
-          .replace(escapedRuleClassName, sanitizeClassName(ruleClassName))
-          .replace(
-            ruleContent,
-            ruleContent
-              .split(";")
-              .map((propertyDeclaration) =>
-                propertyDeclaration.endsWith("!important")
-                  ? propertyDeclaration.trim()
-                  : `${propertyDeclaration.trim()}!important`,
-              )
-              .join(";"),
-          )
-          .replace(/[\r\n|\r|\n]+/g, "") // remove line breaks
-          .replace(/\s+/g, " "),
-      );
+    const newProps = {
+      ...element.props,
+      ...propsToOverwrite,
+    };
+    const newChildren = propsToOverwrite.children
+      ? propsToOverwrite.children
+      : element.props.children;
+
+    if (typeof element.type === "function") {
+      const component = element.type as React.FC;
+      const renderedComponent = component({
+        ...element.props,
+        ...propsToOverwrite,
+      });
+
+      if (React.isValidElement<EmailElementProps>(renderedComponent)) {
+        return processElement(renderedComponent);
+      }
     }
 
-    headStyles.push(finalMediaQuery);
+    return React.cloneElement(element, newProps, newChildren);
   }
 
-  const nonMediaQueryTailwindStylesPerClass =
-    getStylesPerClassMap(nonMediaQueryCSS);
-
-  const childrenArray = React.Children.toArray(children);
-  const validElementsWithIndexes = childrenArray
-    .map((child, i) => [child, i] as [AnyElement, number])
-    .filter(([child]) => React.isValidElement(child));
+  const hasNonInlineStylesToApply = nonInlineStylesToApply.length > 0;
+  let hasAppliedNonInlineStyles = false as boolean;
+
+  const childrenArray =
+    React.Children.map(children, (child) => {
```

Comment:
This is what we use, instead of what we did before — modifying the children through indexes that we have to keep track of in a pretty messy way — which is much simpler and, AFAIK, has not that much of an overhead.